python/packages/autogen-ext/src/autogen_ext/models/cache.py

```diff
@@ -0,0 +1,186 @@
+
+from autogen_core import CancellationToken
+from autogen_core.models import (
+    ChatCompletionClient,
+    CreateResult,
+    LLMMessage,
+    ModelCapabilities,  # type: ignore
+    ModelInfo,
+    RequestUsage,
+)
+from autogen_core.store import AbstractStore
+from autogen_core.tools import Tool, ToolSchema
+
+
+class ChatCompletionCache(ChatCompletionClient):
+    """
+    A wrapper around a ChatCompletionClient that caches creation results from an underlying client.
+    Cache hits do not contribute to token usage of the original client.
+    """
+
+    def __init__(self, client: ChatCompletionClient, store: AbstractStore):
+        """
+        Initialize a new ChatCompletionCache.
+
+        Args:
+            client (ChatCompletionClient): The original ChatCompletionClient to wrap.
+            store (AbstractStore): A store object that implements get and set methods.
+                The user is responsible for managing the store's lifecycle & clearing it (if needed).
+        """
+        self.client = client
+        self.store = store
+
+    def _check_cache(
+        self,
+        messages: Sequence[LLMMessage],
+        tools: Sequence[Tool | ToolSchema],
+        json_output: Optional[bool],
+        extra_create_args: Mapping[str, Any],
+        force_cache: bool,
+        force_client: bool,
+    ) -> tuple[Optional[Union[CreateResult, List[Union[str, CreateResult]]]], str]:
+        """
+        Helper function to check the cache for a result.
+        Returns a tuple of (cached_result, cache_key).
+        cached_result is None if the cache is empty or force_client is True.
+        Raises an error if there is a cache miss and force_cache is True.
+        """
+        if force_client and force_cache:
+            raise ValueError("force_cache and force_client cannot both be True")
+
+        data = {
+            "messages": [message.model_dump() for message in messages],
+            "tools": [(tool.schema if isinstance(tool, Tool) else tool) for tool in tools],
+            "json_output": json_output,
+            "extra_create_args": extra_create_args,
+        }
+        serialized_data = json.dumps(data, sort_keys=True)
+        cache_key = hashlib.sha256(serialized_data.encode()).hexdigest()
+
+        if not force_client:
+            cached_result = cast(Optional[CreateResult], self.store.get(cache_key))
+            if cached_result is not None:
+                return cached_result, cache_key
+            elif force_cache:
+                raise ValueError("Encountered cache miss for force_cache request")
+
+        return None, cache_key
+
+    async def create(
+        self,
+        messages: Sequence[LLMMessage],
+        *,
+        tools: Sequence[Tool | ToolSchema] = [],
+        json_output: Optional[bool] = None,
+        extra_create_args: Mapping[str, Any] = {},
+        cancellation_token: Optional[CancellationToken] = None,
+        force_cache: bool = False,
```

Comment:
Often time the cache client is used just like an ordinary client, so it is unlikely that caller is even aware of the cache. So for now we don't really have usage cases for the `force_cache` or `force_client`. 

However, it would be useful to set the behavior during construction time -- that's when the caller will know what is used. We can allow the inner client to be a `None`, which indicates forced cache behavior. To force client, the caller would just not use the cache.

Reply:
How will the cache be populated if we never have a client? Do you want to add some methods on the cached client to populate `{prompts -> result}` pairs?

Reply:
As discussed offline, removing these awkward forcing APIs for now to keep the initial implementation simple.